packages/block-library/src/heading/edit.js

```diff
@@ -36,6 +38,32 @@ function HeadingEdit( {
 		} ),
 		style: mergedStyle,
 	} );
+	const generatedAnchor = useGeneratedAnchor(
+		clientId,
+		attributes.anchor,
+		content
+	);
+
+	// Update anchor when the content changes.
+	useEffect( () => {
+		if ( generatedAnchor && generatedAnchor.startsWith( 'wp-' ) ) {
+			setAttributes( { anchor: generatedAnchor } );
+		}
+	}, [ generatedAnchor ] );
+
+	// Make sure manually-edited anchors don't have the `wp-` prefix.
+	// `wp-` marks an anchor as auto-generated.
+	useEffect( () => {
+		if (
+			attributes.anchor &&
+			generatedAnchor !== attributes.anchor &&
+			attributes.anchor.startsWith( 'wp-' )
+		) {
+			setAttributes( {
+				anchor: attributes.anchor.replace( 'wp-', '' ),
+			} );
```

Comment:
The really important thing I wanted to mention earlier, and which in the end I forgot:

Normally, in Gutenberg, any setting of attributes happens as a direct consequence of a user action: typing a key in a rich-text field updates `attributes.content`, clicking on _H3_ updates `attributes.level`, etc. This is an important assumption that we implicitly share with the editor and with the user.

Whenever we face a problem that has us automatically set certain attributes, we break that assumption. Here, we have the following chain:

```
user updates heading text   setAttribute({ content })   useEffect's callback        
           or             →           or              →     is called        → setAttribute({ anchor })
user inputs anchor value    setAttribute({ anchor }
```

The most immediate consequence is that **Undo** breaks:

https://user-images.githubusercontent.com/150562/117051859-ea7add00-ad0e-11eb-91dd-ad79f0ec9544.mov

In Gutenberg, this has been coined an _undo trap_ (#8119). We can fix this using `__unstableMarkNextChangeAsNotPersistent` with the same approach as in #26377. That said, we have to be very careful about what else might break when we break the action→attribute assumption.

Reply:
I added `__unstableMarkNextChangeAsNotPersistent` locally to see how it felt, and we see the other consequence of automatic `setAttribute` calls: the fragmentation of the undo history. See the screencast below. When I add a Paragraph, type in a word, and press Undo, the whole word is deleted. With Heading, I need to undo multiple times.


https://user-images.githubusercontent.com/150562/117054753-5ad72d80-ad12-11eb-898c-dae64760ffc6.mov

One way to address this is to delay the effect (e.g. `setTimeout`, `defer`, or `onBlur`, etc.). I don't know if there are better alternatives (cc @ellatrix).

Reply:
I pushed one more commit that fixes this issue... But it's wrong. It works, but there has to be a better way to do it: https://github.com/WordPress/gutenberg/pull/30825/commits/c61d9a75a24461f7d8ce20b65919eb4942b2fe1c :thinking:

Reply:
Issue resolved by refactoring the `useGeneratedAnchor` function, and combining this `setAttributes` with the pre-existing one from the `onChange` event in the input field itself. This also means that `useEffect` was removed, further simplifying the implementation.

Now history works as expected and anchors change along with the input change events. If the anchor was manually edited, changes are detected by the refactored `useGeneratedAnchor` function and it returns the anchor without modifying it.

Reply:
After some more experimentation, I believe it now works as expected... 
History works properly, and anchors get generated without any issues that I can see so far.

https://user-images.githubusercontent.com/588688/117143707-caecbe80-adb9-11eb-8866-548306748d77.mp4